### PR TITLE
support JSON.ARRINDEX command

### PR DESCRIPTION
**Pull request by skyitachi**
#1807

### CODE REVIEW

**Comment by torwig** [comment]
@skyitachi Thank you for your contribution. To avoid lint error during the CI, please run `./x.py format` before the final commit and push.

**Comment by skyitachi** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=apache_kvrocks&pullRequest=1865)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache_kvrocks&pullRequest=1865&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=apache_kvrocks&pullRequest=1865&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_kvrocks&pullRequest=1865&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=apache_kvrocks&pullRequest=1865&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=apache_kvrocks&pullRequest=1865) No Coverage information [](https://sonarcloud.io/component_measures?id=apache_kvrocks&pullRequest=1865&metric=new_duplicated_lines_density&view=list) [4.9% Duplication](https://sonarcloud.io/component_measures?id=apache_kvrocks&pullRequest=1865&metric=new_duplicated_lines_density&view=list)
> 
> idea Catch issues before they fail your Quality Gate with our IDE extension sonarlint [SonarLint](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=sonarcloud-welcome)

it seems no need to care about define const for JSON.ARRINDEX comman in json_test.go files ?

**Comment by PragmaTwice** [comment]
Hi @skyitachi , currently you do not need to care about the sonarcloud report, since we have not finished its set-up yet.

But other CI checks (lint / build / unit tests) are mandatory.

File: src/commands/cmd_json.cc [hunk 1]
```diff
@@ -204,11 +205,59 @@ class CommandJsonArrLen : public Commander {
...
+      }
+    }
+    if (args_.size() > 5) {
+      auto parse_end = ParseInt<int>(args_[5], 10);
+      if (parse_end.IsOK()) {
+        end = parse_end.GetValue();
+      } else {
+        return {Status::RedisParseErr, errValueNotInteger};
+      }
+    }
```
Comment (PragmaTwice) [inline src/commands/cmd_json.cc]: Could you move these code to `Parse` method instead of `Execute`?

Comment (skyitachi) [reply src/commands/cmd_json.cc]: ok, got it

File: src/types/json.h [hunk 2]
```diff
@@ -134,6 +134,34 @@ struct JsonValue {
     return result_count;
   }
 
+  StatusOr<std::vector<ssize_t>> ArrIndex(std::string_view path, const jsoncons::json &needle, int start,
+                                          int end) const {
+    std::vector<ssize_t> result;
+    try {
+      jsoncons::jsonpath::json_query(value, path, [&](const std::string & /*path*/, const jsoncons::json &val) {
+        if (!val.is_array()) {
+          result.emplace_back(-2);
```
Comment (PragmaTwice) [inline src/types/json.h]: Could you declare some constants to state what these magic numbers `-1` and `-2` mean?

File: src/commands/cmd_json.cc [hunk 3]
```diff
@@ -204,11 +205,59 @@ class CommandJsonArrLen : public Commander {
   }
 };
 
+class CommanderJsonArrIndex : public Commander {
+ public:
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    redis::Json json(svr->storage, conn->GetNamespace());
+
+    std::vector<ssize_t> result;
+    int start = -1;
```
Comment (PragmaTwice) [inline src/commands/cmd_json.cc]: Seems here we can use `0` instead of `-1`?

File: src/types/redis_json.h [hunk 4]
```diff
@@ -41,6 +41,8 @@ class Json : public Database {
   rocksdb::Status Clear(const std::string &user_key, const std::string &path, size_t *result);
   rocksdb::Status ArrLen(const std::string &user_key, const std::string &path,
                          std::vector<std::optional<uint64_t>> &arr_lens);
+  rocksdb::Status ArrIndex(const std::string &user_key, const std::string &path, const std::string &needle, int start,
```
Comment (PragmaTwice) [inline src/types/redis_json.h:44]: IMHO it would be better to use `size_t` for the type of parameter `start` and `end`.

File: src/types/json.h [hunk 5]
```diff
@@ -134,6 +134,34 @@ struct JsonValue {
...
+        int pend = std::max(std::min(static_cast<int>(val.size()), end), 0);
+        auto arr_begin = val.array_range().begin();
+        auto begin_it = arr_begin + pstart;
+
+        auto end_it = arr_begin + pend;
+        auto it = std::find(begin_it, end_it, needle);
+        if (it != end_it) {
+          result.emplace_back(it - arr_begin);
+          return;
+        }
```
Comment (PragmaTwice) [inline src/types/json.h]: > Negative values are interpreted as starting from the end.

From the document, I think we need to calculate the real pos while `end` is negative.

Comment (skyitachi) [reply src/types/json.h]: yes, I have rewrite by normalize array indices as [redis_json](https://github.com/RedisJSON/RedisJSON/blob/master/redis_json/src/redisjson.rs#L40C4-L40C4) does.

File: src/commands/cmd_json.cc (hunk 3 above)
Comment (skyitachi) [reply src/commands/cmd_json.cc]: negative start index also need to be normalized, so default value should be 0

File: src/types/json.h (hunk 2 above)
Comment (skyitachi) [reply src/types/json.h]: fixed it

**Comment by PragmaTwice** [comment]
Seems the golangci-lint have failed. You can try to fix your golang code to make it pass.

https://github.com/apache/kvrocks/actions/runs/6734552602/job/18306093590?pr=1865

File: src/commands/cmd_json.cc [hunk 6]
```diff
@@ -291,6 +291,64 @@ class CommandJsonArrPop : public Commander {
...
+    start_ = 0;
+    end_ = std::numeric_limits<ssize_t>::max();
+
+    if (args.size() > 4) {
+      auto parse_start = ParseInt<ssize_t>(args[4], 10);
+      if (parse_start.IsOK()) {
+        start_ = parse_start.GetValue();
+      } else {
+        return {Status::RedisParseErr, errValueNotInteger};
+      }
```
Comment (git-hulk) [inline src/commands/cmd_json.cc]:
```suggestion
      start_ = GET_OR_RET(ParseInt<ssize_t>(args[4], 10));
```

File: src/commands/cmd_json.cc [hunk 7]
```diff
@@ -291,6 +291,64 @@ class CommandJsonArrPop : public Commander {
...
+    }
+    if (args.size() > 5) {
+      auto parse_end = ParseInt<ssize_t>(args[5], 10);
+      if (parse_end.IsOK()) {
+        end_ = parse_end.GetValue();
+      } else {
+        return {Status::RedisParseErr, errValueNotInteger};
+      }
```
Comment (git-hulk) [inline src/commands/cmd_json.cc]:
```suggestion
      end_ = GET_OR_RET(ParseInt<ssize_t>(args[4], 10));
```

File: src/commands/cmd_json.cc [hunk 8]
```diff
@@ -291,6 +291,64 @@ class CommandJsonArrPop : public Commander {
   int64_t index_ = -1;
 };
 
+class CommanderJsonArrIndex : public Commander {
+ public:
+  Status Parse(const std::vector<std::string> &args) override {
+    if (args.size() < 4 || args.size() > 6) {
```
Comment (git-hulk) [inline src/commands/cmd_json.cc]:
```suggestion
    if (args.size() > 6) {
```

Line 361 required the number of arguments must be at least 4, so can remove this check.

**Review by git-hulk — DISMISSED** [review]

File: src/commands/cmd_json.cc (hunk 8 above)
Comment (skyitachi) [reply src/commands/cmd_json.cc]: fixed it， please check

**Comment by git-hulk** [comment]
Hi @skyitachi Would you mind resolving conflicts before merging.

**Comment by git-hulk** [comment]
@skyitachi Seems previous suggestions are disappeared, can you apply them again?

**Comment by skyitachi** [comment]
> @skyitachi Seems previous suggestions are disappeared, can you apply them again?

ok, it's weired and I will apply code change again

**Review by git-hulk — APPROVED** [review]

**Pull request merged**